File: frontend/src/api/progress.ts
import { buildQuery, request } from "./client";
import type { Page } from "./types";
export interface ProgressLevel { id: number; title: string; description: string | null; min_coins: number; is_active: boolean }
export interface CoinProgressData { total: number; available: number; current: ProgressLevel | null; next: ProgressLevel | null; level_number: number; remaining: number; progress: number; levels: ProgressLevel[]; achievements: import("./types").BadgeOut[] }
export interface NotificationEntry { id: number; title: string; body: string; kind: string; link: string | null; read_at: string | null; created_at: string }
export const progressApi = {
  summary: (userId?: number) => request<CoinProgressData>(userId ? `/api/v1/admin/progress/users/${userId}` : "/api/v1/me/progress"),
  levels: () => request<ProgressLevel[]>("/api/v1/admin/progress/levels"),
  saveLevel: (value: Omit<ProgressLevel, "id">, id?: number) => request<ProgressLevel>(`/api/v1/admin/progress/levels${id ? `/${id}` : ""}`, { method: id ? "PUT" : "POST", json: value }),
  notifications: (page: number, unread: boolean) => request<Page<NotificationEntry>>(`/api/v1/me/notifications${buildQuery({ page, size: 20, unread })}`),
  read: (id?: number) => request(`/api/v1/me/notifications${id ? `/${id}` : ""}/read`, { method: "POST" }),
};

File: app/schemas/progress.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.cabinet import BadgeOut
from app.schemas.common import ORMModel


class ProgressLevelOut(ORMModel):
    id: int
    title: str
    min_coins: int
    description: str | None
    is_active: bool


class ProgressLevelIn(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True, extra="forbid")
    title: str = Field(min_length=1, max_length=120)
    min_coins: int = Field(ge=0, le=100000000)
    description: str | None = Field(default=None, max_length=2000)
    is_active: bool = True


class ProgressSummary(BaseModel):
    total: int
    available: int
    current: ProgressLevelOut | None
    next: ProgressLevelOut | None
    level_number: int
    remaining: int
    progress: float
    levels: list[ProgressLevelOut]
    achievements: list[BadgeOut]


class NotificationOut(ORMModel):
    id: int
    title: str
    body: str
    kind: str
    link: str | None
    read_at: datetime | None
    created_at: datetime
